aea/test_tools/test_skill.py

```diff
@@ -0,0 +1,345 @@
+                    return (
+                        False,
+                        "The '{}' fields do not match. Actual '{}': {}. Expected '{}': {}".format(  # pylint: disable=duplicate-string-formatting-argument
+                            attribute_name,
+                            attribute_name,
+                            attribute,
+                            attribute_name,
+                            expected_value,
+                        ),
+                    )
+
+        return True, "The message has the provided expected attributes."
+
+    def build_incoming_message(
+        self,
+        message_type: Type[Message],
+        performative: Message.Performative,
+        dialogue_reference: Optional[Tuple[str, str]] = None,
+        message_id: Optional[int] = None,
+        target: Optional[int] = None,
+        to: Optional[Address] = None,
+        sender: Address = "counterparty",
+        **kwargs,
+    ) -> Message:
+        """
+        Quickly create an incoming message with the provided attributes.
+
+        For any attribute not provided, the corresponding default value in message is used.
+
+        :param message_type: the type of the message
+        :param dialogue_reference: the dialogue_reference
+        :param message_id: the message_id
+        :param target: the target
+        :param performative: the performative
+        :param to: the 'to' address
+        :param sender: the 'sender' address
+        :param kwargs: other attributes
+
+        :return: the created incoming message
+        """
+        message_attributes = dict()  # type: Dict[str, Any]
+
+        default_dialogue_reference = (
+            Dialogues._generate_dialogue_nonce(),  # pylint: disable=protected-access
+            "",
+        )
+        dialogue_reference = (
+            default_dialogue_reference
+            if dialogue_reference is None
+            else dialogue_reference
+        )
+        message_attributes["dialogue_reference"] = dialogue_reference
+        if message_id is not None:
+            message_attributes["message_id"] = message_id
+        if target is not None:
+            message_attributes["target"] = target
+        message_attributes["performative"] = performative
+        message_attributes.update(kwargs)
+
+        incoming_message = message_type(**message_attributes)
+        incoming_message.sender = sender
+        incoming_message.to = (
+            self.skill.skill_context.agent_address if to is None else to
+        )
+
+        return incoming_message
+
+    def build_incoming_message_for_dialogue(
+        self,
+        dialogue: Dialogue,
+        performative: Message.Performative,
+        target: Optional[int] = None,
+        **kwargs,
+    ) -> Message:
+        """
+        Quickly create an incoming message with the provided attributes for a dialogue.
+
+        For any attribute not provided, the corresponding default value in message is used.
```

Comment:
Method to be used with care. The dialogue provided is part of the object which is being tested (the skill). By providing the dialogue and no message_id, target, and dialogue_ref, the test is by designed insured to pass for these three....

Reply:
Done